Sources/SwifterSwift/SwiftStdlib/StringExtensions.swift

```diff
@@ -967,6 +967,17 @@ public extension String {
     }
     #endif
 
+    #if canImport(Foundation)
+    /// SwifterSwift: Verify if string matches the regex.
+    ///
+    /// - Parameter regex: Regex to verify.
+    /// - Returns: true if string matches the regex.
+    func matches(regex: NSRegularExpression) -> Bool {
+        let range = NSRange(location: 0, length: utf16.count)
```

Comment:
Why utf16?

Reply:
I was under the impression that this was the right choice for converting String length to NSRange. I have seen this in loads of places, including SwifterSwift (see `isSpelledCorrectly`). I use this in my app, though we only support a small set of locales.

Is that incorrect?

Reply:
This is more a question, I'm not familiar enough with string encoding/representation, but to me just `count` would always return the correct length of the string and as far as I know strings are backed by a contiguous storage UTF-8 buffer... not really sure if this would make a difference... but again I'm not too familiar with string representation.

Reply:
I've never researched the topic in depth, but now I know the answer! Apple docs to the rescue. When converting to an NSString/NSRange API, utf16.count should be used. See:

https://developer.apple.com/documentation/swift/string/utf16view

I verified this in a playground:

![image](https://user-images.githubusercontent.com/526195/92197057-8f15e980-ee25-11ea-9961-aa15a558b3ec.png)

Reply:
Ah that makes sense! Thanks

Reply:
```suggestion
        let range = NSRange(startIndex..<endIndex, in: self)
```
I really need to add these core `String` extensions, which are the correct way to convert between `Range<String.Index>` and `NSRange`.
```Swift
extension String {
    func range(nsRange: NSRange) -> Range<Index> {
        return Range(nsRange, in: self)!
    }

    func nsRange(range: Range<Index>) -> NSRange {
        return NSRange(range, in: self)
    }

    var fullRange: Range<Index> {
        return startIndex..<endIndex
    }

    var fullNSRange: NSRange {
        return nsRange(range: fullRange)
    }
}
```
The issue with using the `utf16` view is that it used to be correct to use the `utf8` view, but then that was changed (I believe when Swift 4 was released). Don't assume the underlying implementation, even though it's highly unlikely to change.

Reply:
That would be great! I have something similar in my app, though it uses utf16. As you correctly pointed out, these are more robust. Just wish we could use them in the extensions 😄